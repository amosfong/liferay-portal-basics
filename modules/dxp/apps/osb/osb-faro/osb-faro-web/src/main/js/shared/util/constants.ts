import {POSITIONS as ALIGN_POSITIONS} from './align';

export enum AcquisitionTypes {
	Channel = 'CHANNEL',
	Referrer = 'REFERRER',
	SourceMedium = 'SOURCE_MEDIUM'
}

export enum ActivityActions {
	Comments = 3,
	Downloads = 0,
	Previews = 4,
	Submissions = 1,
	Visits = 2
}

export enum Applications {
	Contacts = 'contacts',
	Main = 'main'
}

export enum AssetNames {
	BlogViewed = 'blogViewed',
	CommentPosted = 'commentPosted',
	DocumentDownloaded = 'documentDownloaded',
	DocumentPreviewed = 'documentPreviewed',
	FormSubmitted = 'formSubmitted',
	FormViewed = 'formViewed',
	PageViewed = 'pageViewed',
	WebContentViewed = 'webContentViewed'
}

export enum AssetTypes {
	Asset = 'Asset',
	Blog = 'Blog',
	Document = 'Document',
	Form = 'Form',
	Journal = 'Journal',
	WebContent = 'WebContent',
	WebPage = 'Page'
}

export enum ChannelPermissionTypes {
	AllUsers = 0,
	SelectUsers = 1
}

export enum CompositionTypes {
	AccountInterests = 'accountInterests',
	Acquisitions = 'acquisitions',
	IndividualInterests = 'individualInterests',
	SearchTerms = 'searchTerms',
	SegmentInterests = 'individualSegmentInterests',
	SiteInterests = 'siteInterests'
}

export enum ConjunctionKey {
	And = 'and',
	Or = 'or'
}

export enum CredentialTypes {
	OAuth1 = 'OAuth 1 Authentication',
	OAuth2 = 'OAuth 2 Authentication',
	Token = 'Token Authentication'
}

export enum DataSourceDisplayStatuses {
	Active = 'ACTIVE',
	Configuring = 'CONFIGURING',
	DeleteError = 'DELETE_ERROR',
	InDeletion = 'IN_DELETION',
	Inactive = 'INACTIVE'
}

export enum DataSourceProgressStatuses {
	Completed = 'COMPLETED',
	Failed = 'FAILED',
	InProgress = 'IN_PROGRESS',
	Started = 'STARTED'
}

export enum DataSourceStates {
	ActionNeeded = 'ACTION_NEEDED',
	AnalyticsClientConfigurationFailure = 'ANALYTICS_CLIENT_CONFIGURATION_FAILURE',
	CredentialsInvalid = 'CREDENTIALS_INVALID',
	CredentialsValid = 'CREDENTIALS_VALID',
	Disconnected = 'DISCONNECTED',
	InProgressDeleting = 'IN_PROGRESS_DELETING',
	LiferayVersionInvalid = 'LIFERAY_VERSION_INVALID',
	UndefinedError = 'UNDEFINED_ERROR',
	Ready = 'READY',
	UrlInvalid = 'URL_INVALID',
	Unconfigured = 'UNCONFIGURED'
}
export enum DataSourceStatuses {
	Active = 'ACTIVE',
	Inactive = 'INACTIVE'
}

export enum DataSourceTypes {
	Csv = 'CSV',
	Liferay = 'LIFERAY',
	Salesforce = 'SALESFORCE'
}

export enum EntityTypes {
	Account = 0,
	AccountsSegment = 3,
	Asset = 5,
	DataSource = 1,
	Individual = 2,
	IndividualsSegment = 4,
	Page = 6
}

export enum ExpirationPeriod {
	In30Days = '2592000',
	In6Months = '15778800',
	In1Year = '31557600',
	Indefinite = '3155760000'
}

export enum FaroEnv {
	Local = 'local',
	Production = 'prd',
	Staging = 'stg'
}

export enum FieldContexts {
	Custom = 'custom',
	Demographics = 'demographics',
	Interests = 'interests',
	Organization = 'organization'
}

export enum FieldOwnerTypes {
	Account = 'account',
	Individual = 'individual',
	Organization = 'organization'
}

export enum FieldTypes {
	Boolean = 'Boolean',
	Date = 'Date',
	Number = 'Number',
	String = 'Text'
}

export enum GDPRRequestStatuses {
	Completed = 'COMPLETED',
	Error = 'ERROR',
	Expired = 'EXPIRED',
	Pending = 'PENDING',
	Running = 'RUNNING'
}

export enum GDPRRequestTypes {
	Access = 'ACCESS',
	Delete = 'DELETE',
	Suppress = 'SUPPRESS',
	Unsuppress = 'UNSUPPRESS'
}

export enum JobRunDataPeriods {
	Last7Days = 'LAST_7_DAYS',
	Last30Days = 'LAST_30_DAYS',
	Last180Days = 'LAST_180_DAYS',
	Last365Days = 'LAST_365_DAYS'
}

export enum JobRunFrequencies {
	Every7Days = 'EVERY_7_DAYS',
	Every14Days = 'EVERY_14_DAYS',
	Every30Days = 'EVERY_30_DAYS',
	Manual = 'MANUAL'
}

export enum JobRunStatuses {
	Completed = 'COMPLETED',
	Failed = 'FAILED',
	Published = 'PUBLISHED',
	Running = 'RUNNING'
}

export enum JobStatuses {
	Failed = 'FAILED',
	Pending = 'PENDING',
	Ready = 'READY',
	Running = 'RUNNING',
	Scheduled = 'SCHEDULED'
}

export enum JobTypes {
	ItemSimilarity = 'CONTENT_RECOMMENDATION_ITEM_SIMILARITY'
}

export enum LanguageIds {
	English = 'en_US',
	Japanese = 'ja_JP',
	Portuguese = 'pt_BR',
	Spanish = 'es_ES'
}

export enum OrderByDirections {
	Ascending = 'ASC',
	Descending = 'DESC'
}

export enum PreferencesScopes {
	User = 'user',
	Group = 'group'
}

export enum ProjectStates {
	Activating = 'ACTIVATING',
	AutoRedeployFailed = 'AUTO_REDEPLOY_FAILED',
	Deactivated = 'DEACTIVATED',
	Maintenance = 'MAINTENANCE',
	NotReady = 'NOT READY',
	Ready = 'READY',
	Scheduled = 'SCHEDULED',
	Unavailable = 'UNAVAILABLE',
	Unconfigured = 'UNCONFIGURED'
}

export enum RangeKeyTimeRanges {
	CustomRange = 'CUSTOM',
	Last180Days = '180',
	Last24Hours = '0',
	Last28Days = '28',
	Last30Days = '30',
	Last7Days = '7',
	Last90Days = '90',
	LastYear = '365',
	Yesterday = '1'
}

export enum SegmentStates {
	Disabled = 'DISABLED',
	InProgress = 'IN_PROGRESS',
	Ready = 'READY'
}

export enum SegmentTypes {
	Dynamic = 'DYNAMIC',
	Static = 'STATIC'
}

export enum SessionEntityTypes {
	Account = 'ACCOUNT',
	Individual = 'INDIVIDUAL'
}

export enum Sizes {
	Small = 'sm',
	Medium = 'md',
	Large = 'lg',
	XLarge = 'xl',
	XXLarge = 'xxl',
	XXXLarge = 'xxxl'
}

export enum SubscriptionStatuses {
	Approaching = 1,
	Ok = 0,
	Over = 2
}

export enum TimeIntervals {
	Day = 'day',
	Month = 'month',
	Quarter = 'quarter',
	Week = 'week',
	Year = 'year'
}

export enum TimeSpans {
	AllTime = 'ever',
	LastYear = 'lastYear',
	Last24Hours = 'last24Hours',
	Last7Days = 'last7Days',
	Last28Days = 'last28Days',
	Last30Days = 'last30Days',
	Last90Days = 'last90Days',
	Today = 'today',
	Yesterday = 'yesterday'
}

export enum UserRoleNames {
	Administrator = 'Site Administrator',
	Member = 'Site Member',
	Owner = 'Site Owner'
}

export enum UserStatuses {
	Approved = 0,
	Pending = 1,
	Requested = 2
}

export const ALIGNMENTS_MAP = {
	bottom: ALIGN_POSITIONS.BottomCenter,
	'bottom-left': ALIGN_POSITIONS.BottomLeft,
	'bottom-right': ALIGN_POSITIONS.BottomRight,
	left: ALIGN_POSITIONS.LeftCenter,
	right: ALIGN_POSITIONS.RightCenter,
	top: ALIGN_POSITIONS.TopCenter,
	'top-left': ALIGN_POSITIONS.TopLeft,
	'top-right': ALIGN_POSITIONS.TopRight
};

export const ASSET_METRICS = [
	{
		key: 'abandonmentsMetric',
		selectTitle: `${Liferay.Language.get(
			'form-abandonment'
		)} (${Liferay.Language.get('ratio').toLowerCase()})`,
		title: Liferay.Language.get('form-abandonment'),
		type: 'percentage'
	},
	{
		key: 'clicksMetric',
		selectTitle: `${Liferay.Language.get(
			'asset-clicks'
		)} (${Liferay.Language.get('sum').toLowerCase()})`,
		title: Liferay.Language.get('clicks'),
		type: 'number'
	},
	{
		key: 'completionTimeMetric',
		selectTitle: `${Liferay.Language.get(
			'form-completion-time'
		)} (${Liferay.Language.get('average').toLowerCase()})`,
		title: Liferay.Language.get('completion-time'),
		type: 'time'
	},
	{
		key: 'downloadsMetric',
		selectTitle: `${Liferay.Language.get(
			'asset-downloads'
		)} (${Liferay.Language.get('sum').toLowerCase()})`,
		title: Liferay.Language.get('downloads'),
		type: 'number'
	},
	{
		key: 'readingTimeMetric',
		selectTitle: `${Liferay.Language.get(
			'asset-interaction-time'
		)} (${Liferay.Language.get('average').toLowerCase()})`,
		title: Liferay.Language.get('interaction-time'),
		type: 'time'
	},
	{
		key: 'submissionsMetric',
		selectTitle: `${Liferay.Language.get(
			'form-submissions'
		)} (${Liferay.Language.get('sum').toLowerCase()})`,
		title: Liferay.Language.get('form-submissions'),
		type: 'number'
	},
	{
		key: 'viewsMetric',
		selectTitle: `${Liferay.Language.get(
			'asset-views'
		)} (${Liferay.Language.get('sum').toLowerCase()})`,
		title: Liferay.Language.get('views'),
		type: 'number'
	}
];

export const DATA_RETENTION_PERIOD_KEY = 'data-retention-period';

export const DEVELOPER_MODE = FARO_DEV_MODE;

// LRAC-11571 Disable temporarily Accounts

export const ENABLE_ACCOUNTS = false;

export const ENABLE_ADD_TRIAL_WORKSPACE =
	FARO_ENV === FaroEnv.Local || FARO_ENV === FaroEnv.Staging;

// LRAC - 11652 Hide Displayed Asset Card on the Page report

export const ENABLE_ASSET_CARD = false;

// LRAC-13649 Hide the keywords blocklist screen from Settings>Definitions

export const ENABLE_BLOCKLIST_KEYWORDS = false;

// LRAC-11651 Disable temporarily CSV File

export const ENABLE_CSVFILE = false;

// LRAC-13389 Disable temporarily Delete Property and Delete Data Source buttons

export const ENABLE_DELETE_DATA_SOURCE_BUTTON = false;

export const ENABLE_DELETE_PROPERTY_BUTTON = false;

// LRAC-11650 Hide Form Abandonment Card

export const ENABLE_FORM_ABANDONMENT = false;

// LRAC-10908 Hide global filters

export const ENABLE_GLOBAL_FILTER = false;

// LRAC-13781 [BUG] The Last Seen column only displays the creation date of the token

export const ENABLE_LAST_ACCESS_DATE = false;

// LRAC-11571 Disable temporarily Salesforce

export const ENABLE_SALESFORCE = false;

export const EXPIRATION_DATE_LABELS = {
	[ExpirationPeriod.In30Days]: Liferay.Language.get('30-days'),
	[ExpirationPeriod.In6Months]: Liferay.Language.get('6-months'),
	[ExpirationPeriod.In1Year]: Liferay.Language.get('1-year'),
	[ExpirationPeriod.Indefinite]: Liferay.Language.get('indefinite')
};

export const LANGUAGES = [
	{
		id: LanguageIds.English,
		label: Liferay.Language.get('english')
	},
	{
		id: LanguageIds.Japanese,
		label: Liferay.Language.get('japanese')
	},
	{
		id: LanguageIds.Portuguese,
		label: Liferay.Language.get('portuguese')
	},
	{
		id: LanguageIds.Spanish,
		label: Liferay.Language.get('spanish')
	}
];

export const ONE_DAY = '86400000';

export const ONE_MONTH = '2592000000';

export const POSITIONS = [
	'top',
	'top',
	'right',
	'bottom',
	'bottom',
	'bottom',
	'left',
	'top'
];

export const SEVEN_MONTHS = '18144000000';

export const THIRTEEN_MONTHS = '33696000000';

export const TIME_RANGE_LABELS = {
	[RangeKeyTimeRanges.Last180Days]: Liferay.Language.get('last-180-days'),
	[RangeKeyTimeRanges.Last24Hours]: Liferay.Language.get('last-24-hours'),
	[RangeKeyTimeRanges.Last28Days]: Liferay.Language.get('last-28-days'),
	[RangeKeyTimeRanges.Last30Days]: Liferay.Language.get('last-30-days'),
	[RangeKeyTimeRanges.Last7Days]: Liferay.Language.get('last-seven-days'),
	[RangeKeyTimeRanges.Last90Days]: Liferay.Language.get('last-90-days'),
	[RangeKeyTimeRanges.LastYear]: Liferay.Language.get('last-year'),
	[RangeKeyTimeRanges.Yesterday]: Liferay.Language.get('yesterday')
};

export const TWO_DAYS = '172800000';

const Constants: Window['faroConstants'] = window.faroConstants;

export default Constants;
