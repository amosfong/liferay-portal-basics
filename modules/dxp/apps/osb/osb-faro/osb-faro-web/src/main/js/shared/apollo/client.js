import cache from './cache';
import {ApolloClient} from 'apollo-client';
import {DEVELOPER_MODE} from 'shared/util/constants';
import {get} from 'lodash';
import {HttpLink} from 'apollo-link-http';
import {onError} from 'apollo-link-error';
import {reloadPage} from 'shared/util/router';
import {resolvers} from './resolvers/resolvers';

const groupIdRegex = /^\/workspace\/([a-z0-9._-]+)/;

/**
 * Fecth With Project ID
 * @param {string} uri
 * @param {object} options
 */
const fetchWithGroupId = (uri, options) => {
	const currentUri = new URL(window.location.href);
	const {operationName} = JSON.parse(options.body);
	const matches = currentUri.pathname.match(groupIdRegex);

	if (matches !== null && matches.length > 1) {
		const groupId = matches[1];

		return fetch(
			`${uri}?opname=${operationName}&projectGroupId=${groupId}`,
			options
		);
	}

	return fetch(`${uri}/&opname=${operationName}`, options);
};

const client = new ApolloClient({
	addTypename: true,
	cache,
	connectToDevTools: FARO_DEV_MODE,
	defaultOptions: {
		watchQuery: {
			notifyOnNetworkStatusChange: true
		}
	},
	link: HttpLink.from([
		onError(({operation}) => {
			const status = get(operation.getContext(), ['response', 'status']);

			if (status === 401) {
				reloadPage();
			}
		}),
		new HttpLink({
			credentials: 'same-origin',
			fetch: fetchWithGroupId,
			uri: '/o/cerebro/graphql'
		})
	]),

	resolvers: {
		/**
		 * Queries must render only in the dev mode to avoid
		 * add unnecessary code in the final bundle
		 */
		Query: DEVELOPER_MODE ? resolvers : {}
	}
});

export default client;
