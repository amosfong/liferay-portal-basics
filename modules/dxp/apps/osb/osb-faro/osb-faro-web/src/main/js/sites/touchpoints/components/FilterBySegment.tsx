import * as API from 'shared/api';
import ClayButton from '@clayui/button';
import ClayDropDown from '@clayui/drop-down';
import ClayIcon from '@clayui/icon';
import ClayLabel from '@clayui/label';
import ClayLink from '@clayui/link';
import Loading, {Align} from 'shared/components/Loading';
import NoResultsDisplay from 'shared/components/NoResultsDisplay';
import React, {useMemo, useState} from 'react';
import URLConstants from 'shared/util/url-constants';
import {
	createOrderIOMap,
	getDefaultSortOrder,
	NAME
} from 'shared/util/pagination';
import {getSafeRangeSelectors, getSafeTouchpoint} from 'shared/util/util';
import {RangeSelectors} from 'shared/types';
import {Routes, SEGMENTS, toRoute} from 'shared/util/router';
import {
	SegmentPageViewsQuery,
	SegmentPageViewsQueryData,
	SegmentPageViewsQueryVariables
} from 'shared/queries/SegmentPageViewsQuery';
import {sub} from 'shared/util/lang';
import {useParams} from 'react-router-dom';
import {useQuery} from '@apollo/react-hooks';
import {useQueryPagination} from 'shared/hooks/useQueryPagination';
import {useRequest} from 'shared/hooks/useRequest';

type Item = {
	children: Item[];
	disabled: boolean;
	id: string;
	name: string;
};

interface IFilterBySegment {
	onFilterChange: (item: Item | null) => void;
	rangeSelectors: RangeSelectors;
}

const filterBySegment: React.FC<IFilterBySegment> = ({
	onFilterChange,
	rangeSelectors
}) => {
	const {channelId, groupId, title, touchpoint} = useParams();
	const {delta, orderIOMap, page, query} = useQueryPagination({
		initialOrderIOMap: createOrderIOMap(NAME, getDefaultSortOrder(NAME))
	});
	const [selectedItem, setSelectedItem] = useState(null);

	const {data, loading} = useRequest({
		dataSourceFn: API.individualSegment.search,
		variables: {
			channelId,
			delta,
			groupId,
			orderIOMap,
			page,
			query
		}
	});

	const {data: segmentData, loading: segmentLoading} = useQuery<
		SegmentPageViewsQueryData,
		SegmentPageViewsQueryVariables
	>(SegmentPageViewsQuery, {
		fetchPolicy: 'network-only',
		skip: !data?.items.length,
		variables: {
			canonicalUrl: getSafeTouchpoint(touchpoint),
			channelId,
			segmentIds: data?.items.map(({id}) => id),
			title: decodeURIComponent(title),
			...getSafeRangeSelectors(rangeSelectors)
		}
	});

	const items = useMemo(
		() =>
			data?.items.map(item => {
				const selectedSegmentData = segmentData?.segmentPageViews.find(
					({segmentId}) => segmentId === item.id
				);

				return {
					...item,
					disabled: !selectedSegmentData?.views
				};
			}) ?? [],
		[data, segmentData]
	);

	return (
		<div className='d-flex justify-content-between w-100 analytics-segment-filter-root'>
			<div className='align-items-center d-flex'>
				<Dropdown
					channelId={channelId}
					groupId={groupId}
					items={items}
					loading={loading || segmentLoading}
					onFilterChange={item => {
						setSelectedItem(item);

						onFilterChange(item);
					}}
				/>

				{selectedItem && (
					<ClayLabel
						className='ml-2'
						closeButtonProps={{
							'aria-label': Liferay.Language.get('close'),
							id: 'closeId',
							title: Liferay.Language.get('close')
						}}
						large
						onClick={() => {
							setSelectedItem(null);
							onFilterChange(null);
						}}
					>
						{selectedItem.name}
					</ClayLabel>
				)}
			</div>

			{selectedItem && (
				<div className='d-flex'>
					<ClayButton
						borderless
						data-tooltip
						data-tooltip-align='top'
						displayType='secondary'
						onClick={() => {
							setSelectedItem(null);
							onFilterChange(null);
						}}
						size='sm'
						title={Liferay.Language.get('remove-filter')}
					>
						<ClayIcon symbol='times-circle' />
					</ClayButton>

					<div className='divider' />
				</div>
			)}
		</div>
	);
};

const Dropdown = ({channelId, groupId, items, loading, onFilterChange}) => {
	const [value, setValue] = useState('');

	const filteredItems = useMemo(() => {
		if (!value) {
			return items;
		}

		return items.filter(
			({name}) => name.match(new RegExp(value, 'i')) !== null
		);
	}, [items, value]);

	return (
		<ClayDropDown
			closeOnClick
			trigger={
				<ClayButton
					borderless
					disabled={loading}
					displayType='secondary'
					size='sm'
				>
					{loading && <Loading align={Align.Left} />}

					{Liferay.Language.get('filter')}

					<ClayIcon className='ml-2' symbol='caret-bottom' />
				</ClayButton>
			}
		>
			<ClayDropDown.Search
				onChange={setValue}
				placeholder={Liferay.Language.get('search')}
			/>

			<ClayDropDown.ItemList
				items={[
					{
						children: filteredItems,
						id: 1,
						name: sub(Liferay.Language.get('filter-by-x'), [
							Liferay.Language.get('segment')
						])
					}
				]}
			>
				{(item: Item) => (
					<ClayDropDown.Group
						header={item.name}
						items={item.children}
						key={item.name}
					>
						{(item: Item) => (
							<ClayDropDown.Item
								disabled={item.disabled}
								key={item.name}
								onClick={() => {
									onFilterChange(item);
								}}
							>
								{item.name}
							</ClayDropDown.Item>
						)}
					</ClayDropDown.Group>
				)}
			</ClayDropDown.ItemList>

			{!!items.length && !filteredItems.length && (
				<ClayDropDown.Section>
					<NoResultsDisplay
						description={
							<div
								className='d-flex flex-column justify-content-center'
								style={{minHeight: 240}}
							>
								<div className='h4 no-results-title'>
									{Liferay.Language.get(
										'there-are-no-results-found'
									)}
								</div>

								{Liferay.Language.get(
									'please-try-a-different-search-term'
								)}
							</div>
						}
						title={null}
					/>
				</ClayDropDown.Section>
			)}

			{!items.length && (
				<ClayDropDown.Section>
					<NoResultsDisplay
						description={
							<div
								className='d-flex flex-column justify-content-center'
								style={{minHeight: 240}}
							>
								<div className='h4 no-results-title'>
									{Liferay.Language.get(
										'there-are-no-segments'
									)}
								</div>

								{Liferay.Language.get(
									'start-by-creating-a-segment'
								)}

								<ClayLink
									className='d-block mb-3'
									href={
										URLConstants.SegmentsDocumentationLink
									}
									key='DOCUMENTATION'
									target='_blank'
								>
									{Liferay.Language.get(
										'learn-more-about-segments'
									)}
								</ClayLink>
							</div>
						}
						title={null}
					/>
				</ClayDropDown.Section>
			)}

			<ClayDropDown.Section>
				<ClayLink
					block
					button
					displayType={
						!items.length || !filteredItems.length
							? 'primary'
							: 'secondary'
					}
					href={toRoute(Routes.CONTACTS_LIST_SEGMENT, {
						channelId,
						groupId,
						type: SEGMENTS
					})}
					small
				>
					{Liferay.Language.get('create-segment')}
				</ClayLink>
			</ClayDropDown.Section>
		</ClayDropDown>
	);
};

export default filterBySegment;
