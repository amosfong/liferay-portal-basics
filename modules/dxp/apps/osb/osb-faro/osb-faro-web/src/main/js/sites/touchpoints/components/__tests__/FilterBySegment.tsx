import * as API from 'shared/api';
import client from 'shared/apollo/client';
import FilterBySegment from '../FilterBySegment';
import React from 'react';
import {ApolloProvider} from '@apollo/react-hooks';
import {cleanup, fireEvent, render} from '@testing-library/react';
import {MockedProvider} from '@apollo/react-testing';
import {mockSegmentPageViewsReq} from 'test/graphql-data';
import {RangeKeyTimeRanges} from 'shared/util/constants';
import {StaticRouter} from 'react-router';
import {waitForLoadingToBeRemoved} from 'test/helpers';

jest.unmock('react-dom');

jest.mock('react-router-dom', () => ({
	...jest.requireActual('react-router-dom'),
	useParams: () => ({
		channelId: '456',
		groupId: '123',
		title: 'Liferay DXP - Home',
		touchpoint: 'http://liferay.com'
	})
}));

interface IWrapperComponentProps {
	onFilterChange: (item: any) => void;
	segmentPageViews: {segmentId: string; views: number}[];
}

const WrapperComponent: React.FC<IWrapperComponentProps> = ({
	onFilterChange,
	segmentPageViews
}) => (
	<ApolloProvider client={client}>
		<StaticRouter>
			<MockedProvider
				mocks={[mockSegmentPageViewsReq({segmentPageViews})]}
			>
				<FilterBySegment
					onFilterChange={onFilterChange}
					rangeSelectors={{
						rangeEnd: '',
						rangeKey: RangeKeyTimeRanges.Last24Hours,
						rangeStart: ''
					}}
				/>
			</MockedProvider>
		</StaticRouter>
	</ApolloProvider>
);

const MOCK_SEGMENT = (id: string, name: string) => ({
	activeIndividualCount: 0,
	activitiesCount: 0,
	anonymousIndividualCount: 0,
	channelId: '654287566937865570',
	dateCreated: 1700512862880,
	dateModified: 1700512862880,
	filter:
		"(activities.filterByCount(filter='(activityKey eq ''Page#pageViewed#1fa4f8a263c3ba307844bb9d37ac12e238e9f128aa894c2151e4b7cf5aa5e814'' and day gt ''last24Hours'')',operator='ge',value=1))",
	id,
	includeAnonymousUsers: false,
	individualAddedDate: null,
	individualCount: 1,
	knownIndividualCount: 1,
	lastActivityDate: null,
	name,
	properties: {},
	referencedObjects: {},
	segmentType: 'DYNAMIC',
	state: 'READY',
	status: 'ACTIVE',
	type: 4,
	userName: 'Test Test'
});

describe('FilterBySegment', () => {
	afterEach(cleanup);

	beforeAll(() => {
		jest.useFakeTimers();
	});

	it('should render', async () => {
		const {container, getByText} = render(
			<WrapperComponent
				onFilterChange={jest.fn()}
				segmentPageViews={[
					{
						segmentId: '123',
						views: 100
					}
				]}
			/>
		);

		await waitForLoadingToBeRemoved(container);

		expect(getByText('Create Segment')).toHaveAttribute(
			'class',
			'btn btn-block btn-sm btn-secondary'
		);
		expect(container).toMatchSnapshot();
	});

	it('should open dropdown w/ no segments empty state', async () => {
		// @ts-ignore
		API.individualSegment.search.mockReturnValueOnce(
			Promise.resolve({
				disableSearch: false,
				items: [],
				total: 0
			})
		);

		const {container, getByText} = render(
			<WrapperComponent
				onFilterChange={jest.fn()}
				segmentPageViews={[
					{
						segmentId: '123',
						views: 100
					}
				]}
			/>
		);

		await waitForLoadingToBeRemoved(container);

		expect(getByText('Create Segment')).toHaveAttribute(
			'class',
			'btn btn-block btn-sm btn-primary'
		);

		fireEvent.click(getByText('Filter'));

		expect(getByText('Filter By Segment')).toBeInTheDocument();
		expect(getByText('Create Segment')).toHaveAttribute(
			'href',
			'/workspace/123/456/contacts/segments'
		);
		expect(getByText('There are no segments.')).toBeInTheDocument();
		expect(getByText('Start by creating a segment.')).toBeInTheDocument();
		expect(getByText('Learn more about segments.')).toHaveAttribute(
			'href',
			'https://learn.liferay.com/analytics-cloud/latest/en/people/segments/segments.html'
		);
	});

	it('should open dropdown w/ a list of segments', async () => {
		// @ts-ignore
		API.individualSegment.search.mockReturnValueOnce(
			Promise.resolve({
				disableSearch: false,
				items: [
					MOCK_SEGMENT('123', 'Viewed Page'),
					MOCK_SEGMENT('456', 'Viewed Form'),
					MOCK_SEGMENT('789', 'Viewed Web Content')
				],
				total: 3
			})
		);

		const {container, getByText} = render(
			<WrapperComponent
				onFilterChange={jest.fn()}
				segmentPageViews={[
					{
						segmentId: '123',
						views: 100
					}
				]}
			/>
		);

		await waitForLoadingToBeRemoved(container);

		expect(getByText('Create Segment')).toHaveAttribute(
			'class',
			'btn btn-block btn-sm btn-secondary'
		);

		fireEvent.click(getByText('Filter'));

		expect(getByText('Viewed Page')).toBeInTheDocument();
		expect(getByText('Viewed Form')).toBeInTheDocument();
		expect(getByText('Viewed Web Content')).toBeInTheDocument();
	});

	it('should open dropdown w/ no segments found empty state', async () => {
		// @ts-ignore
		API.individualSegment.search.mockReturnValueOnce(
			Promise.resolve({
				disableSearch: false,
				items: [MOCK_SEGMENT('123', 'Viewed Page')],
				total: 1
			})
		);

		const {container, getByRole, getByText} = render(
			<WrapperComponent
				onFilterChange={jest.fn()}
				segmentPageViews={[
					{
						segmentId: '123',
						views: 100
					}
				]}
			/>
		);

		await waitForLoadingToBeRemoved(container);

		expect(getByText('Create Segment')).toHaveAttribute(
			'class',
			'btn btn-block btn-sm btn-secondary'
		);

		fireEvent.click(getByText('Filter'));

		expect(getByText('Viewed Page')).toBeInTheDocument();

		fireEvent.change(getByRole('textbox'), {
			target: {value: 'Viewed Form'}
		});

		expect(getByText('Create Segment')).toHaveAttribute(
			'class',
			'btn btn-block btn-sm btn-primary'
		);

		expect(getByText('There are no results found.')).toBeInTheDocument();
		expect(
			getByText('Please try a different search term.')
		).toBeInTheDocument();
	});

	it('should open dropdown w/ segments and select one of them', async () => {
		const onFilterChange = jest.fn();

		// @ts-ignore
		API.individualSegment.search.mockReturnValueOnce(
			Promise.resolve({
				disableSearch: false,
				items: [
					MOCK_SEGMENT('123', 'Viewed Page'),
					MOCK_SEGMENT('456', 'Viewed Form'),
					MOCK_SEGMENT('789', 'Viewed Web Content')
				],
				total: 3
			})
		);

		const {container, getByRole, getByText} = render(
			<WrapperComponent
				onFilterChange={onFilterChange}
				segmentPageViews={[
					{
						segmentId: '123',
						views: 100
					},
					{
						segmentId: '456',
						views: 100
					},
					{
						segmentId: '789',
						views: 100
					}
				]}
			/>
		);

		await waitForLoadingToBeRemoved(container);

		expect(getByText('Create Segment')).toHaveAttribute(
			'class',
			'btn btn-block btn-sm btn-secondary'
		);

		fireEvent.click(getByText('Filter'));

		fireEvent.click(getByText('Viewed Page'));

		expect(onFilterChange).toHaveBeenCalledWith(
			expect.objectContaining(MOCK_SEGMENT('123', 'Viewed Page'))
		);

		expect(container.querySelector('.label')).toBeInTheDocument();
		expect(
			getByRole('button', {
				name: /remove filter/i
			})
		).toBeInTheDocument();
	});

	it('should open dropdown w/ segments, select one of them, and then, remove filter', async () => {
		const onFilterChange = jest.fn();

		// @ts-ignore
		API.individualSegment.search.mockReturnValueOnce(
			Promise.resolve({
				disableSearch: false,
				items: [
					MOCK_SEGMENT('123', 'Viewed Page'),
					MOCK_SEGMENT('456', 'Viewed Form'),
					MOCK_SEGMENT('789', 'Viewed Web Content')
				],
				total: 3
			})
		);

		const {container, getByRole, getByText} = render(
			<WrapperComponent
				onFilterChange={onFilterChange}
				segmentPageViews={[
					{
						segmentId: '123',
						views: 100
					},
					{
						segmentId: '456',
						views: 100
					},
					{
						segmentId: '789',
						views: 100
					}
				]}
			/>
		);

		await waitForLoadingToBeRemoved(container);

		expect(getByText('Create Segment')).toHaveAttribute(
			'class',
			'btn btn-block btn-sm btn-secondary'
		);

		fireEvent.click(getByText('Filter'));

		fireEvent.click(getByText('Viewed Page'));

		expect(onFilterChange).toHaveBeenCalledWith(
			expect.objectContaining(MOCK_SEGMENT('123', 'Viewed Page'))
		);

		expect(container.querySelector('.label')).toBeInTheDocument();

		fireEvent.click(
			getByRole('button', {
				name: /remove filter/i
			})
		);

		expect(container.querySelector('.label')).not.toBeInTheDocument();

		expect(onFilterChange).toHaveBeenCalledWith(
			expect.objectContaining([])
		);
	});

	it('should open dropdown w/ segment disabled if there are no views', async () => {
		const onFilterChange = jest.fn();

		// @ts-ignore
		API.individualSegment.search.mockReturnValueOnce(
			Promise.resolve({
				disableSearch: false,
				items: [MOCK_SEGMENT('123', 'Viewed Page')],
				total: 3
			})
		);

		const {container, getByText} = render(
			<WrapperComponent
				onFilterChange={onFilterChange}
				segmentPageViews={[
					{
						segmentId: '123',
						views: 0
					}
				]}
			/>
		);

		await waitForLoadingToBeRemoved(container);

		fireEvent.click(getByText('Filter'));

		const viewedPageSegment = getByText('Viewed Page');

		expect(viewedPageSegment).toHaveAttribute('disabled');

		fireEvent.click(viewedPageSegment);

		expect(onFilterChange).not.toHaveBeenCalledWith();
	});
});
