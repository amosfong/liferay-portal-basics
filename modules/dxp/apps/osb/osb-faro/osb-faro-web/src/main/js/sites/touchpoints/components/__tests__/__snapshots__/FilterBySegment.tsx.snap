// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterBySegment should render 1`] = `
<div>
  <div
    class="d-flex justify-content-between w-100 analytics-segment-filter-root"
  >
    <div
      class="align-items-center d-flex"
    >
      <div
        class="dropdown"
      >
        <button
          aria-controls="clay-dropdown-menu-1"
          aria-expanded="false"
          aria-haspopup="true"
          class="dropdown-toggle btn btn-outline-borderless btn-sm btn-outline-secondary"
          type="button"
        >
          Filter
          <svg
            class="lexicon-icon lexicon-icon-caret-bottom ml-2"
            role="presentation"
          >
            <use
              href="#caret-bottom"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</div>
`;
